lib/marin/src/marin/rl/hooks.py

```diff
@@ -0,0 +1,255 @@
+        """
+        super().__init__(frequency, start_step)
+        self.n_examples = n_examples
+        self.eval_type = eval_type
+        self.evaluate_all_lessons = evaluate_all_lessons
+
+    def run(self, context: HookContext) -> dict[str, Any] | None:
+        """Run evaluation on lessons."""
+        worker = context.worker
+        rng, eval_rng = context.split_rng()
+
+        if self.evaluate_all_lessons:
+            # Full curriculum evaluation
+            logger.info(f"Running full curriculum evaluation at step {context.step}")
+            return worker._evaluate_curriculum(eval_rng, context.step)
+        else:
+            # Single lesson evaluation (micro-eval)
+            if context.lesson_id is None:
+                logger.warning("No lesson_id in context for micro-evaluation")
+                return None
+
+            logger.info(f"Running {self.eval_type} for lesson {context.lesson_id} at step {context.step}")
+            return worker._evaluate_lesson(
+                context.lesson_id, self.n_examples, eval_type=self.eval_type, rng=eval_rng, step=context.step
+            )
+
+    def __repr__(self) -> str:
+        return (
+            f"EvaluationHook(frequency={self.frequency}, n_examples={self.n_examples}, "
+            f"eval_type='{self.eval_type}', evaluate_all_lessons={self.evaluate_all_lessons})"
+        )
+
+
+class HookManager:
+    """Manages the execution of hooks in the RolloutWorker."""
+
+    def __init__(self):
+        """Initialize the hook manager."""
+        self.hooks: list[Hook] = []
+
+    def register_hook(self, hook: Hook) -> None:
+        """Register a new hook.
+
+        Args:
+            hook: The hook to register
+        """
+        self.hooks.append(hook)
+        logger.info(f"Registered hook: {hook}")
+
+    def unregister_hook(self, hook: Hook) -> bool:
+        """Unregister a hook.
+
+        Args:
+            hook: The hook to unregister
+
+        Returns:
+            True if hook was found and removed, False otherwise
+        """
+        if hook in self.hooks:
+            self.hooks.remove(hook)
+            logger.info(f"Unregistered hook: {hook}")
+            return True
+        return False
+
+    def clear_hooks(self) -> None:
+        """Remove all registered hooks."""
+        self.hooks.clear()
+        logger.info("Cleared all hooks")
+
+    def run_hooks(self, context: HookContext) -> dict[str, Any]:
+        """Run all hooks that should execute at the current step.
+
+        Args:
+            context: The hook context
+
+        Returns:
+            Aggregated results from all hooks
+        """
+        results = {}
+        for hook in self.hooks:
+            try:
+                if hook.should_run(context):
+                    logger.debug(f"Running hook: {hook}")
+                    hook_results = hook.run(context)
+                    if hook_results:
+                        results.update(hook_results)
```

Comment:
nit, maybe prefix entries with the hook name:

results.update({ f"{hook.name}/k}: v for (k,v) in hook_results.items() })